pyomo/contrib/alternative_solutions/solnpool.py

```diff
@@ -9,119 +9,1026 @@
+        """
+        return self.active_pool.to_dict()
+
+    def __iter__(self):
+        """
+        Yields
+        -------
+        Solution
+            The solutions in the active pool.
+        """
+        for soln in self.active_pool.solutions:
+            yield soln
+
+    def __len__(self):
+        """
+        Returns
+        -------
+        int
+            The number of solutions in the active pool.
+        """
+        return len(self.active_pool)
+
+    def __getitem__(self, soln_id):
+        """
+        Returns
+        -------
+        Solution
+            The specified solution in the active pool.
+        """
+        return self._pools[self._name][soln_id]
+
+    def add(self, *args, **kwargs):
+        """
+        Adds a solution to the active pool.
+
+        Returns
+        ----------
+        int
+            The index of the solution that is added.
+        """
+        return self.active_pool.add(*args, **kwargs)
+
+    #
+    # The following methods support the management of multiple
+    # pools within a PoolManager.
+    #
+
+    @property
+    def active_pool(self):
+        """
+        Gets the underlying active SolutionPool in PoolManager
+
+        Returns
+        ----------
+        SolutionPool
+            Active pool object
+
+        """
+        assert self._name in self._pools, f"Unknown pool '{self._name}'"
+        return self._pools[self._name]
+
+    def add_pool(self, *, name=None, policy="keep_best", as_solution=None, **kwds):
+        """
+        Initializes a new solution pool and adds it to this pool manager.
+
+        The method expects parameters for the constructor of the corresponding solution pool.
+        Supported pools are `keep_all`, `keep_best`, `keep_latest`, and `keep_latest_unique`.
+
+        Parameters
+        ----------
+        name : String
+            The name of the solution pool.  If name is already used then, then an error is generated.
+        policy : String
+            This string indicates the policy that is enforced new solution pool.
+            Supported values are ['keep_all', 'keep_best', 'keep_latest', 'keep_latest_unique'].
+            (Default is 'keep_best'.)
+        as_solution : Function or None
+            Method for converting inputs into Solution objects.
+            A value of None will result in the default_as_solution function being used.
+        **kwds
+            Other associated arguments that are used to initialize the solution pool.
+
+        Returns
+        ----------
+        dict
+            Metadata for the newly create solution pool.
+        """
+        if name is None and None in self._pools:
+            del self._pools[None]
+
+        if name not in self._pools:
+            # Delete the 'None' pool if it isn't being used
+            if name is not None and None in self._pools and len(self._pools[None]) == 0:
+                del self._pools[None]
+
+            if policy == "keep_all":
```

Comment:
This "if tree" could be replaced by a dispatcher:
```python
class PoolManager:
    _policy_dispatcher = {
        "keep_all": SolutionPool_KeepAll,
        "keep_best": SolutionPool_KeepBest,
        "keep_latest": SolutionPool_KeepLatest,
        "keep_latest_unique": SolutionPool_KeepLatestUnique,
    }
    # ...
    def add_pool(self, *, name=None, policy="keep_best", as_solution=None, **kwds):
        # ...
        self._pool[name] = self._policy_dispatcher[policy](name, as_solution, weakref.proxy(self), **kwds)
```

This both makes the code simpler and also more extensible.

Reply:
Done